File: src/main/java/com/koldyr/csv/processor/export/ExportProcessor.java
package com.koldyr.csv.processor.export;

import java.io.IOException;
import java.sql.Connection;
import java.sql.ResultSet;
import java.sql.ResultSetMetaData;
import java.sql.SQLException;
import java.sql.Statement;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.concurrent.Callable;
import java.util.concurrent.Future;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.koldyr.csv.Constants;
import com.koldyr.csv.io.DBToFilePipeline;
import com.koldyr.csv.model.PageBlockData;
import com.koldyr.csv.model.PoolType;
import com.koldyr.csv.model.ProcessorContext;
import com.koldyr.csv.processor.BatchDBProcessor;

/**
 * Description of class ExportProcessor
 *
 * @created: 2018.03.03
 */
public class ExportProcessor extends BatchDBProcessor {

    private static final Logger LOGGER = LoggerFactory.getLogger(ExportProcessor.class);

    public ExportProcessor(ProcessorContext config) {
        super(config);
    }

    @Override
    protected void processTable(String tableName) {
        long start = System.currentTimeMillis();
        Thread.currentThread().setName(tableName);

        final String fileName = context.getPath() + '/' + tableName + ".csv";

        Connection connection = null;
        try (DBToFilePipeline dataPipeline = new DBToFilePipeline(fileName)) {
            connection = context.get(PoolType.SOURCE);
            long rowCount = getRowCount(connection, tableName);

            LOGGER.debug("Starting table {}: {} rows", tableName, format.format(rowCount));

            if (rowCount > context.getPageSize()) {
                release(connection, PoolType.SOURCE);
                connection = null; //don't release this connection in finally block

                parallelExport(dataPipeline, tableName, rowCount);
            } else {
                export(connection, dataPipeline, tableName, rowCount);
            }

            LOGGER.debug("Finished table {}: {} rows in {} ms", tableName, format.format(rowCount), format.format(System.currentTimeMillis() - start));
        } catch (Exception e) {
            LOGGER.error(e.getMessage(), e);
        } finally {
            release(connection, PoolType.SOURCE);
        }
    }

    private void export(Connection connection, DBToFilePipeline dataPipeline, String tableName, double rowCount) throws SQLException, IOException {
        final double step = context.getPageSize() / 100.0;

        ResultSet resultSet = null;
        try (Statement statement = connection.createStatement()) {
            resultSet = statement.executeQuery("SELECT * FROM \"" + context.getSrcSchema() + "\".\"" + tableName + "\"");

            final ResultSetMetaData metaData = resultSet.getMetaData();
            final int columnCount = metaData.getColumnCount();

            int counter = 0;
            while (dataPipeline.next(resultSet, columnCount)) {
                counter++;

                if (counter % step == 0) {
                    dataPipeline.flush();
                    final long percent = Math.round(counter / rowCount * 100.0);
                    LOGGER.debug("\t{}%", percent);
                }
            }
        } finally {
            if (resultSet != null) {
                resultSet.close();
            }
        }
    }

    private void parallelExport(DBToFilePipeline dataPipeline, String tableName, long rowCount) throws InterruptedException {
        int pageCount = (int) Math.ceil(rowCount / (double) context.getPageSize());

        LOGGER.debug("Export {} pages", pageCount);

        List<PageBlockData> pages = new ArrayList<>(pageCount);

        for (int i = 0; i < pageCount; i++) {
            long start = i * context.getPageSize();
            long length = Math.min(context.getPageSize(), rowCount - start);
            pages.add(new PageBlockData(i, start, length));
        }

        context.setPages(tableName, pages);

        int threadCount = Math.min(Constants.PARALLEL_PAGES, pageCount);
        final Collection<Callable<Integer>> exportThreads = new ArrayList<>(threadCount);
        for (int i = 0; i < threadCount; i++) {
            exportThreads.add(new PageExportProcessor(context, tableName, dataPipeline));
        }

        final List<Future<Integer>> results = context.getExecutor().invokeAll(exportThreads);

        checkResults(tableName, pageCount, results);
    }
}

File: src/main/kotlin/com/koldyr/csv/processor/export/ExportProcessor.kt
package com.koldyr.csv.processor.export

import com.koldyr.csv.Constants.FETCH_SIZE
import com.koldyr.csv.Constants.PARALLEL_PAGES
import com.koldyr.csv.db.SQLStatementFactory
import com.koldyr.csv.io.DBToFilePipeline
import com.koldyr.csv.model.PageBlockData
import com.koldyr.csv.model.PoolType
import com.koldyr.csv.model.ProcessorContext
import com.koldyr.csv.processor.BatchDBProcessor
import org.slf4j.LoggerFactory
import java.io.IOException
import java.nio.file.Path
import java.sql.Connection
import java.sql.SQLException
import java.util.concurrent.Callable
import kotlin.math.ceil
import kotlin.math.min
import kotlin.math.roundToLong

/**
 * Description of class ExportProcessor
 *
 * @created: 2018.03.03
 */
class ExportProcessor(context: ProcessorContext) : BatchDBProcessor(context) {

    override fun processTable(tableName: String) {
        val start = System.currentTimeMillis()
        Thread.currentThread().name = tableName

        val csvFile = Path.of(context.path, "$tableName.csv")

        var connection: Connection? = null
        try {
            DBToFilePipeline(csvFile).use { dataPipeline ->
                connection = context[PoolType.SOURCE]
                val rowCount = getRowCount(connection!!, tableName)

                LOGGER.debug("Starting table {}: {} rows", tableName, format.format(rowCount))

                if (rowCount > context.pageSize) {
                    release(connection, PoolType.SOURCE)
                    connection = null //don't release this connection in finally block

                    parallelExport(dataPipeline, tableName, rowCount)
                } else {
                    export(connection!!, dataPipeline, tableName, rowCount.toDouble())
                }

                if (LOGGER.isDebugEnabled) {
                    val count = format.format(rowCount)
                    val duration = format.format(System.currentTimeMillis() - start)
                    LOGGER.debug("Finished table {}: {} rows in {} ms", tableName, count, duration)
                }
            }
        } catch (e: Exception) {
            LOGGER.error(e.message, e)
        } finally {
            release(connection, PoolType.SOURCE)
        }
    }

    @Throws(SQLException::class, IOException::class)
    private fun export(connection: Connection, dataPipeline: DBToFilePipeline, tableName: String, rowCount: Double) {
        val step = context.pageSize / 100.0

        connection.createStatement().use { statement ->
            val selectAll = SQLStatementFactory.getSelectAll(connection, context.srcSchema, tableName)
            statement.fetchSize = min(FETCH_SIZE.toDouble(), rowCount).toInt()

            statement.executeQuery(selectAll).use { resultSet ->
                val metaData = resultSet.metaData
                val columnCount = metaData.columnCount

                var counter = 0
                while (dataPipeline.next(resultSet, columnCount)) {
                    counter++

                    if (counter % step == 0.0) {
                        dataPipeline.flush()
                        val percent = (counter / rowCount * 100.0).roundToLong()
                        LOGGER.debug("\t{}%", percent)
                    }
                }
            }
        }
    }

    @Throws(InterruptedException::class)
    private fun parallelExport(dataPipeline: DBToFilePipeline, tableName: String, rowCount: Long) {
        val pageCount = ceil(rowCount / context.pageSize.toDouble()).toInt()

        LOGGER.debug("Export {} pages", pageCount)

        val pages = ArrayList<PageBlockData>(pageCount)

        for (i in 0 until pageCount) {
            val start = i * context.pageSize
            val length = min(context.pageSize, rowCount - start)
            pages.add(PageBlockData(i, start, length))
        }

        context.setPages(tableName, pages)

        val threadCount = min(PARALLEL_PAGES, pageCount)
        val exportThreads = ArrayList<Callable<Int>>(threadCount)
        for (i in 0 until threadCount) {
            exportThreads.add(PageExportProcessor(context, tableName, dataPipeline))
        }

        val results = context.executor.invokeAll(exportThreads)

        checkResults(tableName, pageCount, results)
    }

    companion object {
        private val LOGGER = LoggerFactory.getLogger(ExportProcessor::class.java)
    }
}
